lib/Dialect/FIRRTL/Transforms/MergeConnections.cpp

```diff
@@ -0,0 +1,254 @@
+          .template Case<SubfieldOp>([&](SubfieldOp subfield) {
+            checkSourceAggregate(subfield, idx, subfield.fieldIndex());
+          })
+          .template Case<SubindexOp>([&](SubindexOp subindex) {
+            checkSourceAggregate(subindex, idx, subindex.index());
+          })
+          .Default([&](auto) { canUseSourceAggregate = false; });
+    }
+
+    // If it is fine to use `sourceAggregate` as a merged value, we just
+    // return it.
+    if (canUseSourceAggregate) {
+      LLVM_DEBUG(llvm::dbgs() << "Success to merge " << fieldRef.getValue()
+                              << " ,fieldID= " << fieldRef.getFieldID() << "to"
+                              << sourceAggregate << "\n";);
+      return sourceAggregate;
+    }
+
+    // Otherwise, we concat all values and cast them into the aggregate type.
+    Value accumulate;
+    changed = true;
+    for (auto value : operands) {
+      value = builder.createOrFold<BitCastOp>(
+          value.getLoc(),
+          UIntType::get(value.getContext(),
+                        *firrtl::getBitWidth(
+                            value.getType().template cast<FIRRTLType>())),
+          value);
+      accumulate = (accumulate ? builder.createOrFold<CatPrimOp>(
+                                     accumulate.getLoc(), value, accumulate)
+                               : value);
+    }
+    return builder.createOrFold<BitCastOp>(accumulate.getLoc(), type,
+                                           accumulate);
+  };
+
+  if (auto bundle = type.dyn_cast_or_null<BundleType>())
+    return getMergedValue(bundle);
+  if (auto vector = type.dyn_cast_or_null<FVectorType>())
+    return getMergedValue(vector);
+
+  return Value();
+}
+
+void MergeConnection::peelConnect(ConnectOp connect) {
+  // Ignore connections between different types because it will produce a
+  // partial connect.
+  if (connect.src().getType() != connect.dest().getType())
+    return;
+
+  auto destFieldRef = getFieldRefFromValue(connect.dest());
+  auto destRoot = destFieldRef.getValue();
+
+  // If dest is derived from mem op or has a ground type, we cannot merge them.
+  if (destRoot.getDefiningOp<MemOp>() ||
+      destRoot.getType().cast<FIRRTLType>().isGround())
+    return;
+
+  fieldRefToConnect[destFieldRef] = connect;
+  peeledRoot.insert(destRoot);
+  return;
+}
+
+bool MergeConnection::run() {
+  // Peel all connections.
+  moduleOp.walk([&](ConnectOp connect) { peelConnect(connect); });
+
+  // Iterate over roots.
+  for (auto root : peeledRoot) {
+    // If the root has a ground type, there is no connection to merge.
+    if (root.getType().cast<FIRRTLType>().isGround())
+      continue;
+
+    unsigned size = shouldBeRemoved.size();
+    auto value = getMergedConnectedValue({root, 0});
+    if (!value) {
+      // If we failed to get merged value, revert dead connections.
+      shouldBeRemoved.resize(size);
+      continue;
+    }
+    auto builder = OpBuilder::atBlockEnd(moduleOp.getBody());
```

Comment:
This pass assumes expandwhens is done?

Reply:
Yes, definitely. Without expandWhens, this pass is illegal.